app/src/main/java/tech/ula/utils/FilesystemUtility.kt

```diff
@@ -87,4 +100,74 @@ class FilesystemUtility(
             }
         }
     }
+
+    fun backupFilesystemByLocation(execDirectory: String, directoryName: String, backupName: String, destinationDir: File) {
+        val backupLocation = "../${backupName}"
+        val backupLocationTmp = "${backupLocation}.tmp"
+        val exclude = "--exclude=sys --exclude=dev --exclude=proc --exclude=data --exclude=mnt --exclude=host-rootfs --exclude=sdcard --exclude=etc/mtab --exclude=etc/ld.so.preload"
+        val commandCompress = "rm -rf ${backupLocationTmp} && tar ${exclude} -cvpzf ${backupLocationTmp} ../${directoryName}"
+        val commandMove = "rm -rf ${applicationFilesDirPath}/${backupName} && mv ${backupLocationTmp} ${backupLocation}"
+        // TODO add in progress notification
+        execUtility.wrapWithBusyboxAndExecute("${execDirectory}", commandCompress, doWait = true)
+        execUtility.wrapWithBusyboxAndExecute("${execDirectory}", commandMove, doWait = true)
+        val backupLocationF = File("${applicationFilesDirPath}/${backupName}")
+        // TODO refactor to not using str.replace
+        val formatter = SimpleDateFormat("yyyy_MM_dd_HH_mm_ss", Locale.US)
+        val now = Date()
+        val fileName = "." + formatter.format(now) + ".tar.gz"
+        backupLocationF.copyTo(File(destinationDir, backupName.replace(".tar.gz",fileName)), overwrite = true)
+    }
+
+    // https://developertip.wordpress.com/2012/12/11/android-copy-file-programmatically-from-its-uri/
+    fun copyFileFromUri(context: Context, saveFilePath: String, fileUri: Uri): Boolean {
```

Comment:
To expand on this a bit: if we have metadata about where the backups are this should become as easy as `backupFile.copyTo()`

Reply:
I did not find any better way to copy file from URI. I was forced to use URI because it was what Intent result returned. Even tough backups will be tracked by app itself there should be mechanism allowing users to load backup from outside of application scope (cloud, drive or other thrid party space) and I think it may use URI anyway (I do not know of any other way how to load file from g.e. google drive yet). I suggest I will add button "Load backup" into future BackupListFragment that would load/download backup from external source in same manner (and adding metadata record about backup to internal handler - db table). App would not rely on external source of backups but it would be optional source of backups considering some feature requests and issues regarding running out of space for multiple filesystems on some devices.

DB sketch would be:
```
{
        "tableName": "backup",
        "createSql": "CREATE TABLE IF NOT EXISTS `${TABLE_NAME}` (`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `name` TEXT NOT NULL, `location` TEXT NOT NULL, `filesystemId` INTEGER NOT NULL",
        "fields": [
          {
            "fieldPath": "id",
            "columnName": "id",
            "affinity": "INTEGER",
            "notNull": true
          },
          {
            "fieldPath": "name",
            "columnName": "name",
            "affinity": "TEXT",
            "notNull": true
          },
          {
            "fieldPath": "location",
            "columnName": "location",
            "affinity": "TEXT",
            "notNull": true
          },
          {
            "fieldPath": "filesystemId",
            "columnName": "filesystemId",
            "affinity": "INTEGER",
            "notNull": true
          }
        ],
        "primaryKey": {
          "columnNames": [
            "id"
          ],
          "autoGenerate": true
        },
        "indices": [
          {
            "name": "index_backup_name",
            "unique": true,
            "columnNames": [
              "name"
            ],
            "createSql": "CREATE UNIQUE INDEX `index_backup_name` ON `${TABLE_NAME}` (`name`)"
          }
        ],
        "foreignKeys": [
          {
            "table": "filesystem",
            "onDelete": "CASCADE",
            "onUpdate": "NO ACTION",
            "columns": [
              "filesystemId"
            ],
            "referencedColumns": [
              "id"
            ]
          }
        ]
      }
``` 

where field 'name' would have default value "<filesystemName>-<architecture>-<dateCreated>" and it could be changed in UI, field 'location' may be file path or URI distinguished by prefix/protocol - idk yet.

Reply:
The issue with the URI intent you're looking for right now is that it's looking for data of any type. We should be looking for a much more specific type of data, something like `application/ulaBackup` or `application/*`.  What happened when you tried using `application/gzip`?